templates/commands/cel.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/abcxyz/abc/templates/model"
+	"github.com/abcxyz/pkg/logging"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+var (
+	celRegistry = types.NewEmptyRegistry()
+
+	celFuncs = []cel.EnvOption{
+		// Adds a split method on strings. Example: "foo,bar".split(",")
+		//
+		// Design decision: use a method instead of a plain function, because
+		// that is the convention that the community seems to prefer for string
+		// splitting. This is based on some quick googling of other peoples'
+		// custom split functions.
+		cel.Function(
+			"split",
+			cel.MemberOverload(
+				"string_split",
+				[]*cel.Type{cel.StringType, cel.StringType},
+				types.NewListType(types.StringType),
+				cel.BinaryBinding(func(lhs, rhs ref.Val) ref.Val {
+					toSplit := lhs.Value().(string)
+					splitOn := rhs.Value().(string)
+					tokens := strings.Split(toSplit, splitOn)
+					return types.NewStringList(celRegistry, tokens)
+				}),
+			),
+		),
+	}
+)
+
+// celCompile parses and compiles the given expr into executable Program.
+func celCompile(ctx context.Context, scope *scope, expr model.String) (cel.Program, error) {
+	startedAt := time.Now()
+
+	celOpts := []cel.EnvOption{}
+	for varName := range scope.All() {
+		celOpts = append(celOpts, cel.Variable(varName, cel.StringType))
+	}
+	celOpts = append(celOpts, celFuncs...) // Add custom function bindings
+
+	env, err := cel.NewEnv(celOpts...)
+	if err != nil {
+		return nil, expr.Pos.AnnotateErr(fmt.Errorf("internal error: failed configuring CEL environment: %w", err))
+	}
+
+	ast, issues := env.Compile(expr.Val)
+	if err := issues.Err(); err != nil {
+		return nil, expr.Pos.AnnotateErr(fmt.Errorf("failed compiling CEL expression: %w", err))
+	}
+
+	prog, err := env.Program(ast)
+	if err != nil {
+		return nil, expr.Pos.AnnotateErr(fmt.Errorf("failed constructing CEL program: %w", err))
+	}
+
+	logger := logging.FromContext(ctx).Named("celCompile")
+	latency := time.Since(startedAt)
+	logger.Debugw("CEL compilation and loading time", "duration_usec", latency.Microseconds())
```

Comment:
One of the nice things about structured logging is that we can provide both machine and human friendly values. It might be nice to have:

```suggestion
	logger.Debugw("CEL compilation and loading time", 
		"duration_usec", latency.Microseconds(),
		"duration_human", latency.String())
```

Reply:
That is nice, thanks. Done.